### ISSUE TITLE
HTML5 elements for twisted.web.template

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5306]
|[<img alt="MostAwesomeDude's avatar" src="https://avatars.githubusercontent.com/u/118035?s=50" width="50" height="50">](https://github.com/MostAwesomeDude)| @MostAwesomeDude reported|
|-|-|
|Trac ID|trac#5306|
|Type|enhancement|
|Created|2011-10-11 04:12:32Z|
|Branch|https://github.com/twisted/twisted/tree/html5-5306|

While moving some HTML5-savvy code to Twisted Web, I discovered that a few fancy new tags were missing from the tag list, so I added them. Patch attached.

Attachments:

* [patch-5306.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/8de/8de8d2b3d001d4a9feed2e7cfb4e74ec81703c19/6d3f83b869cc69d14cc136bad074d3390354cec9.diff) (2129 bytes) - added by MostAwesomeDude on 2011-10-11 04:13:58Z - Patch adding HTML5 tags to t.web.template
* [patch-5306.2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/8de/8de8d2b3d001d4a9feed2e7cfb4e74ec81703c19/860d48519262d3fcd2e2ed6a3fea26f405ef345f.diff) (4845 bytes) - added by MostAwesomeDude on 2011-10-15 18:25:47Z - Updated patch

<details><summary>Searchable metadata</summary>

```
trac-id__5306 5306
type__enhancement enhancement
reporter__MostAwesomeDude MostAwesomeDude
priority__normal normal
milestone__ 
branch__branches_html5_5306 branches/html5-5306
branch_author__MostAwesomeDude MostAwesomeDude
status__closed closed
resolution__fixed fixed
component__web web
keywords__patch patch
time__1318306352000000 1318306352000000
changetime__1318739219000000 1318739219000000
version__None None
owner__exarkun exarkun
cc__jknight cc__mithrandi
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="MostAwesomeDude's avatar" src="https://avatars.githubusercontent.com/u/118035?s=50" width="50" height="50">](https://github.com/MostAwesomeDude)<a name="note_2"></a>|@MostAwesomeDude commented|
|-|-|

Opening for review. Given the original reviews for this swath of code, I'm relatively sure that no extra unit tests are needed, just a careful once-over.

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_3"></a>|@mithrandi set owner to @MostAwesomeDude|
|-|-|

I apologize in advance for the ridiculously long review given how trivial this patch is, but hopefully all of my suggestions are trivial to implement.

 1. I would suggest removing (or rather, not adding) the `<xmp>` element; the HTML 5 spec documents this under "non-conforming features", saying that "[it is] entirely obsolete, and must not be used by authors", and if you actually need the obsolete magic parsing semantics of `<xmp>`, you won't actually be able to output the string you want using `t.w.t`'s element construction APIs.
 1. All changes to Twisted need a news fragment; I don't think it's necessary to list all of the added elements, but perhaps `<audio>` and `<video>` deserve an honorable mention.
 1. I assume you referred to a specific portion of the HTML 5 spec in order to compile the new list of elements; it would be helpful to future helionauts if you included a link to the relevant portion in a comment. Additionally, if you implement point 1 above, I'd suggest briefly mentioning that `<xmp>` has been excluded from the list deliberately.

Once the above points have been addressed, I believe this change is ready to be merged. I'm not sure if you're a committer or not; please reassign to me if you are not, and I will commit it to trunk.

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_4"></a>|@mithrandi commented|
|-|-|

Since I already did the work of generating the list of added element names (no element names were removed), I thought I'd put it here in case it's useful to anyone else:

article, aside, audio, canvas, command, datalist, details, embed, figcaption,
figure, footer, header, hgroup, keygen, mark, meter, nav, output, progress, rp,
rt, ruby, section, source, summary, time, video, wbr, xmp

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

> I'm relatively sure that no extra unit tests are needed, just a careful once-over.

He he he.  Good one.

It looks like there is little or no meaningful coverage of the attribute restriction logic on `twisted.web.template.tags`.  The coverage tool may report that the `AttributeError` case is exercised, but that is only because trial accidentally triggers it during test suite collection.

**Comment by twisted-trac** [comment]
|[<img alt="MostAwesomeDude's avatar" src="https://avatars.githubusercontent.com/u/118035?s=50" width="50" height="50">](https://github.com/MostAwesomeDude)<a name="note_6"></a>|@MostAwesomeDude commented|
|-|-|

Hi,

Attaching a new patch. Differences:

 1. <xmp> is removed.
 1. <bdi> is added.
 1. A new test case has been added, which exercises general tag lookup, failed tag lookup, HTML5 tag lookup, transparent tag lookup, and <xmp> tag lookup. This brings t.w.template coverage to 100% and adequately exercises every possible path through _TagFactory.__getattr__().
 1. A topfile has been added. I'm claiming this is a feature. <canvas> and <video> get mentioned.
 1. A comment explaining when the tags were last updated, and linking to the W3Schools page listing HTML5 tags, has been added. The comment also mentions <xmp>.

I went through the tag list, more exhaustively, and I found two things. One is that <bdi> was missing; I added that. The other is <isindex>. It's massively deprecated and never actually was standard. Since we also don't have <marquee>, should <isindex> be removed?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [32848]) Branching to 'html5-5306'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @exarkun|
|-|-|

Thanks.  This looks good to me.  I'll merge it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set status to `closed`|
|-|-|

(In [32865]) Merge html5-5306

Author: MostAwesomeDude
Reviewer: exarkun
Fixes: [#5306](https://github.com/twisted/twisted/issues/5306)

Add `audio` and `video` tags to `twisted.web.template.tags` and generally try to be
HTML5 compliant in the supported tag set.
